pkg/rekor/v2/client.go

```diff
@@ -18,29 +18,105 @@ import (
 	"context"
 	"fmt"
 	"net/url"
+	"slices"
 	"time"
 
-	prototrustroot "github.com/sigstore/protobuf-specs/gen/pb-go/trustroot/v1"
 	tiles_client "github.com/sigstore/rekor-tiles/pkg/client"
 	"github.com/sigstore/rekor-tiles/pkg/client/read"
 	"github.com/sigstore/sigstore-go/pkg/root"
+	"github.com/sigstore/sigstore-go/pkg/tuf"
 	"github.com/sigstore/sigstore/pkg/signature"
 )
 
 type ShardInfo struct {
-	client   *read.Client
-	verifier *signature.Verifier
+	client      *read.Client
+	verifier    *signature.Verifier
+	validityEnd time.Time
 }
 
-func GetRekorShards(ctx context.Context, trustedRoot *root.TrustedRoot, rekorServices []root.Service, userAgent string) (map[string]ShardInfo, string, error) {
-	rekorServices, err := root.SelectServices(rekorServices, root.ServiceConfiguration{Selector: prototrustroot.ServiceSelector_ALL}, []uint32{2}, time.Now())
+func RefreshSigningConfig(tufClient *tuf.Client) (*root.SigningConfig, error) {
+	err := tufClient.Refresh()
+	if err != nil {
+		return nil, fmt.Errorf("error refreshing TUF client: %v", err)
+	}
+
+	signingConfig, err := root.GetSigningConfig(tufClient)
 	if err != nil {
-		return nil, "", fmt.Errorf("error selecting rekor services: %v", err)
+		return nil, fmt.Errorf("error getting SigningConfig target: %v", err)
+	}
+	return signingConfig, nil
+}
+
+func filterV2Shards(rekorServices []root.Service) []root.Service {
+	// First we sort and filter the Rekor services so that they're ordered from
+	// newest to oldest and only include the v2 logs
+	sortedServices := make([]root.Service, len(rekorServices))
+	copy(sortedServices, rekorServices)
+	slices.SortFunc(sortedServices, func(i, j root.Service) int {
+		return i.ValidityPeriodStart.Compare(j.ValidityPeriodStart)
+	})
+	slices.Reverse(sortedServices)
+
+	var rekorV2Services []root.Service
+	for _, s := range sortedServices {
+		if s.MajorAPIVersion == 2 {
+			rekorV2Services = append(rekorV2Services, s)
+		}
+	}
+
+	return rekorV2Services
+}
+
+func ShardsNeedUpdating(currentShards map[string]ShardInfo, newSigningConfig *root.SigningConfig) (bool, error) {
```

Comment:
could we run `GetRekorShards` on the new signing config and do a `reflect.DeepEqual(old, new)` instead? Or is this different in some subtle way?

Reply:
`GetRekorShards` does some extra network requests in order to populate the `ShardInfo` struct. It creates a new client per shard, fetches the latest checkpoint for each, and creates a verifier for each.

Since `ShardsNeedUpdating` will run on every iteration, I made it so that it only compares the metadata in `SigningConfig` without any extra network requests.